build/tasks/compare_size.mjs

```diff
@@ -0,0 +1,132 @@
+import fs from "fs";
+import { promisify } from "util";
+import zlib from "node:zlib";
+import { exec as nodeExec } from "child_process";
+
+const gzip = promisify( zlib.gzip );
+const exec = promisify( nodeExec );
+
+async function getBranchName() {
+	const { stdout } = await exec( "git rev-parse --abbrev-ref HEAD" );
+	return stdout.trim();
+}
+
+async function getCache( loc ) {
+	try {
+		const contents = await fs.promises.readFile( loc, "utf8" );
+		return JSON.parse( contents );
+	} catch ( err ) {
+		return {};
+	}
+}
+
+function saveCache( loc, cache ) {
+	return fs.promises.writeFile( loc, JSON.stringify( cache ) );
+}
+
+async function cleanWorkingDir() {
+	const { stdout } = await exec( "git status --untracked-files=no --porcelain" );
+	return !stdout.trim();
+}
+
+function compareSizes( existing, current, padLength ) {
+	if ( typeof current !== "number" ) {
+		return chalk.grey( `${existing}`.padStart( padLength ) );
+	}
+	const delta = current - existing;
+	if ( delta > 0 ) {
+		return chalk.red( `+${delta}`.padStart( padLength ) );
+	}
+	return chalk.green( `${delta}`.padStart( padLength ) );
+}
+
+async function compareSize( { cache = ".sizecache.json", files } = {} ) {
+	if ( !files || !files.length ) {
+		throw new Error( "No files specified" );
+	}
+
+	const branch = await getBranchName();
+	const sizeCache = await getCache( cache );
+
+	let rawPadLength = 0;
+	let gzPadLength = 0;
+	const results = await Promise.all(
+		files.map( async function( filename ) {
+			const contents = await fs.promises.readFile( filename, "utf8" );
+			const size = Buffer.byteLength( contents, "utf8" );
+			const gzippedSize = ( await gzip( contents ) ).length;
+
+			// Add one to give space for the `+` or `-` in the comparison
+			rawPadLength = Math.max( rawPadLength, size.toString().length + 1 );
+			gzPadLength = Math.max( gzPadLength, gzippedSize.toString().length + 1 );
+
+			return { filename, raw: size, gz: gzippedSize };
+		} )
+	);
+
+	const header = "raw".padStart( rawPadLength ) +
+		"gz".padStart( gzPadLength + 1 ) +
+		" Filename";
+
+	const output = [
+		"", // Opening new line
+		chalk.bold( "Sizes" ),
+		header,
+		...results.map( function( result ) {
+			return (
+				result.raw.toString().padStart( rawPadLength ) +
+				" " +
+				result.gz.toString().padStart( gzPadLength ) +
+				" " +
+				result.filename
+			);
+		} ),
+		...Object.keys( sizeCache ).map( function( branch ) {
+			return "\n" + chalk.bold( branch ) + "\n" + header + "\n" +
+				Object.keys( sizeCache[ branch ] ).map( function( filename ) {
+					const result = sizeCache[ branch ][ filename ];
+					const compareResult = results.find( function( result ) {
+						return result.filename === filename;
+					} ) || {};
+					return (
+						compareSizes( result.raw, compareResult.raw, rawPadLength ) +
+						" " +
+						compareSizes( result.gz, compareResult.gz, gzPadLength ) +
+						" " +
+						filename
+					);
+				} ).join( "\n" );
+		} ),
```

Comment:
I think it'd be easier to read if you extracted some of the inner computations to separate variables.

Also, I'd prefer to use template literals instead of concatenation.

Reply:
I also normally prefer template literals, unless the line is very long like this one. Then you're only option for new lines is within the `${}`, which I don't think reads as easily. For instance, the lines above might look like this...

```js
return `${compareSizes( result.raw, compareResult.raw, rawPadLength )} ${
	compareSizes( result.gz, compareResult.gz, gzPadLength )
} ${filename}`;
```

And I don't think that looks as each expression getting its own line. It also takes longer to tell how much space is between each one.

Reply:
Ok, but if each of those expressions is in a variable, the one line wouldn't be too long.